hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/payload/ExpressionPayload.scala

```diff
@@ -164,6 +165,47 @@ class ExpressionPayload(@transient record: GenericRecord,
     }
   }
 
+  private def doRecordMerge(incomingRecord: GenericRecord,
+                            existingRecord: IndexedRecord,
+                            schema: Schema,
+                            properties: Properties): HOption[IndexedRecord] = {
+    val tablePayloadClass = properties.getProperty(PAYLOAD_ORIGINAL_AVRO_PAYLOAD)
+    if (tablePayloadClass.equals(classOf[OverwriteWithLatestAvroPayload].getName)) {
+      HOption.of(incomingRecord)
+    } else if (tablePayloadClass.equals(classOf[DefaultHoodieRecordPayload].getName)) {
+      if (needUpdatingPersistedRecord(existingRecord.asInstanceOf[GenericRecord], incomingRecord, properties)) {
+        HOption.of(incomingRecord)
+      } else {
+        HOption.of(existingRecord)
+      }
+    } else {
+      val orderingField = ConfigUtils.getOrderingField(properties)
+      if (StringUtils.isNullOrEmpty(orderingField)) {
+        HOption.of(incomingRecord)
+      } else {
+        val consistentLogicalTimestampEnabled = properties.getProperty(KeyGeneratorOptions.KEYGENERATOR_CONSISTENT_LOGICAL_TIMESTAMP_ENABLED.key,
+          KeyGeneratorOptions.KEYGENERATOR_CONSISTENT_LOGICAL_TIMESTAMP_ENABLED.defaultValue).toBoolean
+        val incomingRecordPayload = HoodieRecordUtils.loadPayload(tablePayloadClass, incomingRecord,
+          HoodieAvroUtils.getNestedFieldVal(incomingRecord, orderingField, true, consistentLogicalTimestampEnabled)
+            .asInstanceOf[Comparable[_]]).asInstanceOf[HoodieRecordPayload[_ <: HoodieRecordPayload[_]]]
```

Comment:
for CUSTOM merge mode, we need to load the payload class whose construction is based on incoming record, so i don't think we can do this upfront with only record payload API. Perhaps, we can use record merge API but it's a larger refactoring.